satpy/composites/__init__.py

```diff
@@ -585,43 +588,99 @@ def __call__(self, projectables, **kwargs):
         coszen = coszen.clip(0, 1)
 
-        # Apply enhancements to get images
-        day_data = enhance2dataset(day_data)
-        night_data = enhance2dataset(night_data)
+        if self.chose_day_night == "day_and_night":
+            # This means both day and night part will be kept
+            data_2 = projectables[1]
+            day_data = data_1
+            night_data = data_2
 
-        # Adjust bands so that they match
-        # L/RGB -> RGB/RGB
-        # LA/RGB -> RGBA/RGBA
-        # RGB/RGBA -> RGBA/RGBA
-        day_data = add_bands(day_data, night_data['bands'])
-        night_data = add_bands(night_data, day_data['bands'])
+            # Apply enhancements to get images
+            day_data = enhance2dataset(day_data)
+            night_data = enhance2dataset(night_data)
 
-        # Replace missing channel data with zeros
-        day_data = zero_missing_data(day_data, night_data)
-        night_data = zero_missing_data(night_data, day_data)
+            # Adjust bands so that they match
+            # L/RGB -> RGB/RGB
+            # LA/RGB -> RGBA/RGBA
+            # RGB/RGBA -> RGBA/RGBA
+            day_data = add_bands(day_data, night_data['bands'])
+            night_data = add_bands(night_data, day_data['bands'])
 
-        # Get merged metadata
-        attrs = combine_metadata(day_data, night_data)
+            # Replace missing channel data with zeros
+            day_data = zero_missing_data(day_data, night_data)
+            night_data = zero_missing_data(night_data, day_data)
+
+            # Get merged metadata
+            attrs = combine_metadata(day_data, night_data)
+
+            # Blend the two images together
+            data = (1 - coszen) * night_data + coszen * day_data
+            data.attrs = attrs
+
+            # Split to separate bands so the mode is correct
+            data = [data.sel(bands=b) for b in data['bands']]
+
+        elif self.chose_day_night == "only_day":
+            # This means only day part will be kept
+            day_data = data_1
+
+            # Apply enhancements to get images
+            day_data = enhance2dataset(day_data)
+
+            # Add alpha band
+            # L -> LA
+            # RGB -> RGBA
+            day_data = add_bands(day_data, day_data['bands'])
+
+            # Replace missing channel data with zeros
+            day_data = zero_missing_data(day_data, day_data)
+
+            # Get merged metadata
+            attrs = combine_metadata(day_data, day_data)
+
+            # Set night part to 0 and get the maskout image
+            data = (1 - coszen) * 0 + coszen * day_data
+            data.attrs = attrs
+
+            # Split to separate bands so the mode is correct
+            data = [data.sel(bands=b) for b in data['bands']]
 
-        # Blend the two images together
-        data = (1 - coszen) * night_data + coszen * day_data
-        data.attrs = attrs
+        elif self.chose_day_night == "only_night":
```

Comment:
Can the day/night only modes be combined with if statements for the things that are different? You could also ignore the `0` parameters in the logic so that extra computation isn't done when it isn't needed. Oh you might be able to combine all three day/night modes...

```python
day_data = projectables[0] if "day" in self.day_night else None
night_data = projectables[0] if self.day_night == "night_only" else None
night_data = projectables[1] if self.day_night == "day_night" else night_data

day_portion = coszen * (day_data if day_data is not None else 0)
night_portion = (1 - coszen) * (night_data if night_data is not None else 0)
data = night_portion + day_portion

```

You'd need some more logic for handling the `add_bands` and `zero_missing_data` stuff, but it may actually make the code a lot smaller. Maybe not clearer though... :fearful:

Reply:
Yeah I also think there's something need to be improved in the logic structure. I'll go look into this.